### PR TITLE
Added Coding Conventions to a docs folder

### PR DESCRIPTION
**Pull request by charleskawczynski**
Any suggestions for modifying this?

cc/ @tapios @lcw @jkozdon @christophernhill

### CODE REVIEW

**Comment by christophernhill** [comment]
@charleskawczynski - these look generally good. Some comments....

I am not sure what I know
```
2 or more, use a "for" - If you are looping over indexes, and you loop over at least two, then just use a for loop.
```
means. Would it be possible to provide a _do this_ and _not this_ type example?

I agree with @jkozdon that putting Julia link first would be good.

I suspect Unicode recommendation may be setting up a losing battle!!

File: docs/CodingConventions.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    - https://docs.julialang.org/en/v1/manual/style-guide/index.html
+
+
+## * Why no Unicode for now?
+
+ - Some characters are visibly indistinguishable. Capital "a" and capital alpha are visibly indistinguishable, but are recognized as separate characters (e.g., search distinguishable).
+
+ - Some characters are difficult to read. Sometimes, the overline/overdot/hats overlap with characters making them difficult to see.
+
+ - Portability issues. Unicode does not render in Jupyter notebook natively (on OSX).
```
[the remark this answers is not in the document]
Comment (charleskawczynski) [reply docs/CodingConventions.md]: Oh, okay. Good to know!

[the remark this answers is not in the document]
Comment (charleskawczynski) [reply docs/CodingConventions.md]: Wait, I've seen Unicode render on github, though. So, how is this a browser issue?

**Comment by tapios** [comment]
Thanks for putting this together. I'd add two other suggestions from the roadmap on Overleaf:

1. Make names consistent, distinctive, and meaningful.  For variable names, follow CMIP conventions (http://clipc-services.ceda.ac.uk/dreq/index.htm) where possible and practicable. [The reason for following CMIP where possible is that we will need to compute a lot of model diagnostics in the end, for the data assimilation part, and if there is substantial overlap between our variable names and CMIP5 names (which are used in observing system simulators), it will make the interfaces more transparent.]

2. Document design and purpose, rather than mechanics and implementation (document interfaces and embed documentation in code).

**Comment by charleskawczynski** [comment]
>  For variable names, follow CMIP conventions (http://clipc-services.ceda.ac.uk/dreq/index.htm) where possible and practicable.

@tapios That link is giving me a 404 not found. Is this [link](http://clipc-services.ceda.ac.uk/dreq/) okay? Or is there a specific one of these you had in mind?

**Comment by tapios** [comment]
> > For variable names, follow CMIP conventions (http://clipc-services.ceda.ac.uk/dreq/index.htm) where possible and practicable.
> 
> @tapios That link is giving me a 404 not found. Is this [link](http://clipc-services.ceda.ac.uk/dreq/) okay?

Yes (trailing l in html got dropped in cutting and pasting...)

**Comment by jkozdon** [comment]
> for variable names, follow CMIP conventions (http://clipc-services.ceda.ac.uk/dreq/index.htm) where possible and practicable.

For us dumb mathematicians this would be super useful I think.

**Comment by charleskawczynski** [comment]
I do have a question. If struct names cannot collide with module names, and we need to use TitleCase for module names, how will we name a module and class for something like, e.g., a grid.

`module Grid`

`struct Grid`
`  n_points :: Float64`
`  ...`
`end`

`end`

doesn't work (module and struct names must be unique). My original thought was that we could follow Python's naming convention with, e.g.,

`module grid`

`struct Grid`
`  n_points :: Float64`
`  ...`
`end`

`end`

But this is not the Julia way. Does someone have a suggestion on what to do here? I like the idea of keeping things modular and easily being able to find where a particular class is..

**Comment by charleskawczynski** [comment]
> Personally I'd pick two different names rather than using a different style. Make one of the names more specific perhaps. What type of grid? Cartesian finite difference? Unstructured mesh? Maybe the module could have a different name, `Grids`, `GridTypes`, etc..

I think that specific struct names do work in some cases, but it can cause names to get too long also. I do like the idea of changing between singular and plural. This also implies that there may be more than one struct within the same module. Would it be pigeonholing us if I added this recommendation to the list?

**Comment by glwagner** [comment]
The julia documentation provides a [recommendation for package names that addresses this question](https://docs.julialang.org/en/v0.6/manual/packages/#Guidelines-for-naming-a-package-1):

> 3. Packages that provide most of their functionality in association with a new type should have pluralized names.

So, we should expect a package called "Grids.jl" to implement a type or abstract type called `Grid`.

**Comment by charleskawczynski** [comment]
> The julia documentation provides a [recommendation for package names that addresses this question](https://docs.julialang.org/en/v0.6/manual/packages/#Guidelines-for-naming-a-package-1):
> 
> > 1. Packages that provide most of their functionality in association with a new type should have pluralized names.
> 
> So, we should expect a package called "Grids.jl" to implement a type or abstract type called `Grid`.

That looks great, I'll add this link to the list as well since it's under a different section of the Julia docs.

**Pull request merged**